go/pkg/casng/upload_digester.go

```diff
@@ -129,7 +138,220 @@ func (u *uploader) digester() {
+		}()
+	}
+
+	stats := Stats{}
+	var err error
+	deferredWg := make(map[string]*sync.WaitGroup)
+	walker.DepthFirst(req.Path, req.Exclude, walker.Callback{
+		Err: func(path impath.Absolute, realPath impath.Absolute, errVisit error) bool {
+			log.V(3).Infof("[casng] upload.digester.visit.err; path=%s, real_path=%s, err=%s, req=%s, tag=%s, walk=%s", path, realPath, errVisit, req.id, req.tag, walkID)
+			err = errors.Join(errVisit, err)
+			return false
+		},
+		Pre: func(path impath.Absolute, realPath impath.Absolute) (walker.PreAction, bool) {
+			log.V(3).Infof("[casng] upload.digester.visit.pre; path=%s, real_path=%s, req=%s, tag=%s, walk=%s", path, realPath, req.id, req.tag, walkID)
+
+			select {
+			// If the request was aborted, abort.
+			case <-req.ctx.Done():
+				log.V(3).Infof("upload.digester.req.cancel; tag=%s, walk=%s", req.tag, walkID)
+				return walker.SkipPath, false
+			// This is intended to short-circuit a cancelled uploader. It is not intended to abort by cancelling the uploader's context.
+			case <-u.ctx.Done():
+				log.V(3).Infof("upload.digester.cancel; tag=%s, walk=%s", req.tag, walkID)
+				return walker.SkipPath, false
+			default:
+			}
+
+			// A cache hit here indicates a cyclic symlink with the same requester or multiple requesters attempting to upload the exact same path with an identical filter.
+			// In both cases, deferring is the right call. Once the requset is processed, all requestters will revisit the path to get the digestion result.
+			// If the path was not cached before, claim it by makring it as in-flight.
+			key := path.String() + req.Exclude.String()
+			wg := &sync.WaitGroup{}
+			wg.Add(1)
+			m, ok := u.nodeCache.LoadOrStore(key, wg)
+			if !ok {
+				// Claimed. Access it.
+				log.V(3).Infof("[casng] upload.digester.visit.claimed; path=%s, real_path=%s, req=%s, tag=%s, walk=%s", path, realPath, req.id, req.tag, walkID)
+				return walker.Access, true
+			}
+
+			// Defer if in-flight. Wait if already deferred since there is nothing else to do.
+			if wg, ok := m.(*sync.WaitGroup); ok {
+				if deferredWg[key] == nil {
+					deferredWg[key] = wg
+					log.V(3).Infof("[casng] upload.digester.visit.defer; path=%s, real_path=%s, req=%s, tag=%s, walk=%s", path, realPath, req.id, req.tag, walkID)
+					return walker.Defer, true
+				}
+				log.V(3).Infof("[casng] upload.digester.visit.wait; path=%s, real_path=%s, req=%s, tag=%s, walk=%s", path, realPath, req.id, req.tag, walkID)
+				startTime := time.Now()
+				wg.Wait()
+				log.V(3).Infof("[casng] upload.digester.visit.wait.duration; start=%d, end=%d, path=%s, real_path=%s, req=%s, tag=%s, walk=%s", startTime.UnixNano(), time.Now().UnixNano(), path, realPath, req.id, req.tag, walkID)
+				delete(deferredWg, key)
+				return walker.Defer, true
+			}
+
+			node := m.(proto.Message) // Guaranteed assertion because the cache is an internal field.
```

Comment:
optional: is it possible to avoid reusing the cache for different value types? It should make it easier to grasp and make more maintainable

Reply:
I assume you are referring to `sync.WaitGroup` and `proto.Message`. I opted to reusing the cache given it already has a generic type for the value. For readers, it's either the desired value, a empty marker, or a reserved marker. In the last two cases, the reader can reserve the key or try to read again later, respectively.

It's possible to have a separate cache for reservations, but I think it's going to increase the amount of code and memory since there would be two nil checks and two placeholders: one for the reservation cache to say this is computed, and another in the node cache for the actual value. This would also introduce an edge case (internal bug) where the reservation cache says it's computed but the node cache says it's nil.

I think it might be worth exploring further as a possible improvement in another PR.